test/end2end_test.go

```diff
@@ -6880,6 +6880,170 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
+	net.Conn
+	CloseCh *testutils.Channel
+}
+
+// Close closes the connection and sends a value on the close channel.
+func (cw *connWrapperWithCloseCh) Close() error {
+	err := cw.Conn.Close()
+	cw.CloseCh.Replace(nil)
+	return err
+}
+
+// This custom creds are used for storing the connections made by the client.
+// The CloseCh in conn can be used to detect when conn is closed.
+type transportRestartCheckCreds struct {
+	mu          sync.Mutex
+	connections []connWrapperWithCloseCh
+}
+
+func (c *transportRestartCheckCreds) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, nil, nil
+}
+func (c *transportRestartCheckCreds) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	conn := &connWrapperWithCloseCh{Conn: rawConn, CloseCh: testutils.NewChannel()}
+	c.connections = append(c.connections, *conn)
+	authInfo := &authInfoWithConn{CommonAuthInfo: credentials.CommonAuthInfo{SecurityLevel: credentials.NoSecurity}, connection: conn}
+	return conn, authInfo, nil
+}
+func (c *transportRestartCheckCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+func (c *transportRestartCheckCreds) Clone() credentials.TransportCredentials {
+	return &transportRestartCheckCreds{}
+}
+func (c *transportRestartCheckCreds) OverrideServerName(s string) error {
+	return nil
+}
+
+type streamWithPeer struct {
+	stream testpb.TestService_FullDuplexCallClient
+	peer   *peer.Peer
+	id     string
+	conn   net.Conn
+}
+
+func (s *streamWithPeer) sendAndReceive() error {
+	if err := s.stream.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		return fmt.Errorf("sending on %s: %v", s.id, err)
+	}
+	if _, err := s.stream.Recv(); err != nil {
+		return fmt.Errorf("receiving on %s: %v", s.id, err)
+	}
+	return nil
+}
+
+// Tests that the client transport drains and restarts when next stream ID exceeds
+// MaxStreamID. This test also verifies that subsequent RPCs use a new client
+// transport and the old transport is closed.
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for {
+				if _, err := stream.Recv(); err != nil {
+					return err
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return err
+				}
+			}
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
Use `defaultTestTimeout` instead of hard-coding to `10s`?

Reply:
forgot this one.. updating this